[Qiu Qianhui] iP

QQH0828 opened the pull request:
# Duke

> “Your mind is for having ideas, not holding them.” – David Allen ([Source](https://dansilvestre.com/productivity-quotes))

Duke is a chatbot that frees your mind from having to remember the things you need to do. It's

- Text-Based
- Easy to learn
- ~~Fast~~ _Super Fast_ to use

All you need to do is, 👍 

1. Download it from [here](https://github.com/QQH0828/ip/releases/tag/v0.1).
2. Double-click it
3. Add your tasks
4. Let it manage your tasks for you 


And it is ****FREE****! :partying_face:


Features: 👍 

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders
- [ ] GUI interface

---

If you are a random programmer that's looking through this, you can use it to practice Java as well. Here is the `main` code for the launcher.

``` js
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

QQH0828 committed:
Add Increment Level-1

QQH0828 committed:
Add Increment Level-2

QQH0828 committed:
Add Increment Level-3 with a new class Task

QQH0828 committed:
Add Increments Level-4

QQH0828 committed:
Add Increments for Automated Text UI Testing

QQH0828 committed:
Add Increment Level-5

QQH0828 committed:
Add Increment Level-6

ShanHng commented on src/main/java/Duke.java:
```diff
@@ -45,22 +67,51 @@ public void addTask(Task task) {
+            throw new DukeException(("☹ OOPS!!! The due date of a deadline cannot be empty.\n"));
+        }
+
+        String info[] = command.split(" /by ");
```
Perhaps we should put the [] right after the String instead even though this is valid too?

QQH0828 replied:
Thanks for pointing them out!

ShanHng commented on src/main/java/Duke.java:
```diff
@@ -45,22 +67,51 @@ public void addTask(Task task) {
+
+        String info[] = command.split(" /by ");
         String description = info[0];
         String by = info[1];
```
Maybe instead of just "by", we can use a variable name that indicates that this is a date?

ShanHng commented on src/main/java/Duke.java:
```diff
@@ -45,22 +67,51 @@ public void addTask(Task task) {
+            throw new DukeException("\t ☹ OOPS!!! The description, from date or due date of a event cannot be empty.\n");
+        }
+
+        String info[] = command.split(" /from ");
```
Perhaps the naming of this variable can be a standard plural noun instead?

ShanHng commented on src/main/java/Duke.java:
```diff
@@ -14,43 +14,88 @@ public void start() {
+    }
+
+    public void event(String command) {
+        String info[] = command.split(" ", 2)[1].split(" /from ");
```
Ooh I like your clever usage of split to distinguish between /to argument and /from argument!

ShanHng commented on src/main/java/Duke.java:
```diff
@@ -14,43 +14,88 @@ public void start() {
+    public void addTask(Task task) {
+        tasks[taskIndex] = task;
+        taskIndex++;
+        System.out.println("\t Got it. I've added this task:\n"
```
I like what you did here - indenting and spacing a sentence into parts that are more readable!

ShanHng commented on src/main/java/Duke.java:
```diff
@@ -14,43 +14,88 @@ public void start() {
+    }
+
+    public void deadline(String command) {
+        String info[] = command.split(" ", 2)[1].split(" /by ");
```
Perhaps we could use a variable name that indicates this is an array? Also, remember to have the square brackets right after the data type~! (String[])

ShanHng reviewed (commented):
Code is clean and easy to read! But there is repeated violation of the rule to have the [ ] placed behind the data type instead of variable name. Generally looks great!

QQH0828 committed:
Add Increment Level -7 as parallel branche with Save method

QQH0828 committed:
Merge branch 'branch-Level-7'

QQH0828 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Event.java
#             src/main/java/Deadline.java

QQH0828 committed:
Add Increments as parallel branche-Level-8 with Dates and Times

QQH0828 committed:
Add Increments as parallel branch-Level-7 with Save method

QQH0828 committed:
Merge branch 'branch-Level-7'

QQH0828 committed:
Merge branch 'branch-Level-8'

QQH0828 committed:
Solve merge conflict branch-Level-8

QQH0828 committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java
#	src/main/java/Task.java

QQH0828 committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S2/ip

QQH0828 committed:
Add Increment: A-MoreOOP

QQH0828 committed:
Add Increment: A-Packages

QQH0828 committed:
Add Increment: A-Gradle

QQH0828 committed:
Add Increment: A-JUnit

QQH0828 committed:
Add Increments as parallel branches: A-JavaDoc

QQH0828 committed:
Add Increments as parallel branches: Level-9

QQH0828 committed:
Add Increments as parallel branches: Level-9

QQH0828 committed:
Add Increments as parallel branches:  A-CodingStandard

QQH0828 committed:
Merge branch 'branch-A-JavaDoc'

QQH0828 committed:
Merge branch 'branch-Level-9'

QQH0828 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/command/AddDeadlineCommand.java
#	src/main/java/command/AddEventCommand.java
#	src/main/java/command/AddTaskCommand.java
#	src/main/java/command/AddTodoCommand.java
#	src/main/java/command/Command.java
#	src/main/java/command/DeleteTaskCommand.java
#	src/main/java/command/ExitCommand.java
#	src/main/java/command/ListCommand.java
#	src/main/java/command/MarkTaskCommand.java
#	src/main/java/storage/Storage.java
#	src/main/java/userinteraction/Ui.java

QQH0828 committed:
Add Increments as branches: A-CheckStyle

QQH0828 committed:
Delete unused classes

QQH0828 committed:
Add Increments as branches: Level-10

QQH0828 committed:
Merge branch 'branch-Level-10'

QQH0828 committed:
Add assert statement in the TaskList class

Assertions are used to define assumptions about the program state
so that the runtime can verify them.

Adding assertions in the code statements so that it can check whether
the split operations used in the input string result in the correct
output.

It can help to detect the mistakes in the codes.

QQH0828 committed:
Modify the codes based on the code quality

Some of the classes had some code quality violation that needed to
be fixed according to the code quality rules.

Deleting dead code and unused codes so that not trip up readers.

Avoiding deep nesting and complicated expressions so that much more
easier for reading.

Trying to follow the single level of abstraction principle.

These are helping to run-time efficiency and understandability.

QQH0828 committed:
Merge pull request #2 from QQH0828/branch-A-Assertions

Add assert statement in the TaskList class

QQH0828 committed:
Merge branch 'master' of https://github.com/QQH0828/ip

* 'master' of https://github.com/QQH0828/ip:
  Add assert statement in the TaskList class

QQH0828 committed:
Merge pull request #3 from QQH0828/branch-A-CodeQuality

Modify the codes based on the code quality

QQH0828 committed:
Merge branch 'master' of https://github.com/QQH0828/ip

* 'master' of https://github.com/QQH0828/ip:
  Modify the codes based on the code quality

QQH0828 committed:
Merge branch 'master' into branch-A-Assertions

QQH0828 committed:
Merge branch 'master' into branch-A-CodeQuality

QQH0828 committed:
Merge branch 'branch-A-Assertions'

QQH0828 committed:
Merge branch 'branch-A-CodeQuality'

QQH0828 committed:
Add Increment: A-CI

QQH0828 committed:
Add an extension: C-Help

Adding help command in the Duke to display all available commands
with the explanation about how to use each commands.

QQH0828 committed:
Merge branch 'branch-C-Help'

QQH0828 committed:
Fix CheckStyle errors

Add a new line in each classes.
Remove whitespace of "trailing whitespace".
Add a space before "{".
Add and change some java doc.

QQH0828 committed:
Add a unique personality

Mainly from gui part.

QQH0828 committed:
Resolve the command error

Solve the issues about spacing after some commands.

QQH0828 committed:
Fix the import format order issues

QQH0828 committed:
Fix import format in parser class

QQH0828 committed:
Fix checkstyle error

QQH0828 committed:
Add java doc in ui class

QQH0828 committed:
Fix checkstyle error of DeadlineTest and DateTimeTest classes

QQH0828 committed:
Fix the checkstyle error in the DateTimeTest class

QQH0828 committed:
Set up a product website with a representative screenshot and user guide

QQH0828 committed:
Add help command in the user guide

QQH0828 committed:
Finalize the features

QQH0828 committed:
Fix no match tasks while using find command

QQH0828 committed:
Fix some errors to make it more user friendly
